core/storage/pager.rs

```diff
@@ -1048,26 +1050,49 @@ impl Pager {
                 _attempts += 1;
             }
         }
-        self.wal_checkpoint(wal_checkpoint_disabled)?;
+        self.wal_checkpoint(wal_checkpoint_disabled, CheckpointMode::Passive)?;
         Ok(())
     }
 
     #[instrument(skip_all, level = Level::DEBUG)]
-    pub fn wal_checkpoint(&self, wal_checkpoint_disabled: bool) -> Result<CheckpointResult> {
+    pub fn wal_checkpoint(
+        &self,
+        wal_checkpoint_disabled: bool,
+        mode: CheckpointMode,
+    ) -> Result<CheckpointResult> {
         if wal_checkpoint_disabled {
-            return Ok(CheckpointResult {
-                num_wal_frames: 0,
-                num_checkpointed_frames: 0,
-            });
+            return Ok(CheckpointResult::default());
         }
 
-        let checkpoint_result = self.io.block(|| {
+        let mut checkpoint_result = self.io.block(|| {
             self.wal
                 .borrow_mut()
-                .checkpoint(self, Rc::new(RefCell::new(0)), CheckpointMode::Passive)
+                .checkpoint(self, Rc::new(RefCell::new(0)), mode)
```

Comment:
Why we create new write counter `Rc::new(RefCell::new(0))` on every iteration inside `io.block` fn?